kratos/mpi/utilities/mpi_search_utilities.h

```diff
@@ -0,0 +1,131 @@
+//                   Vicente Mataix Ferrandiz
+
+#pragma once
+
+// System includes
+#include <vector>
+
+// External includes
+
+// Project includes
+#include "containers/array_1d.h"
+
+namespace Kratos
+{
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/**
+ * @class MPISearchUtilities
+ * @ingroup KratosMPI
+ * @brief MPI utilities for searching geometrical objects
+ * @details Original implementation from MappingUtilities 
+ * @author Philipp Bucher (moved by Vicente Mataix Ferrandiz)
+ */
+class MPISearchUtilities
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// The Bounding Box type
+    using BoundingBoxType = std::array<double, 6>;
+
+    /// The index type definition
+    using IndexType = std::size_t;
+
+    /// The size type definition
+    using SizeType = std::size_t;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief Check if a point is inside a bounding box
+     * @param rBoundingBox The bounding box
+     * @param rCoords The point
+     * @return true if the point is inside the bounding box
+     */
+    static bool PointIsInsideBoundingBox(
+        const BoundingBoxType& rBoundingBox,
+        const array_1d<double, 3>& rCoords
+        )
+    {
+        // The Bounding Box should have some tolerance already!
+        if (rCoords[0] < rBoundingBox[0] && rCoords[0] > rBoundingBox[1])   // check x-direction
+            if (rCoords[1] < rBoundingBox[2] && rCoords[1] > rBoundingBox[3])   // check y-direction
+                if (rCoords[2] < rBoundingBox[4] && rCoords[2] > rBoundingBox[5])   // check z-direction
+                    return true;
+        return false;
+    }
+
+    /**
+     * @brief Compute the bounding boxes of the given bounding boxes from a given tolerance
+     * @param rBoundingBoxes The bounding boxes
+     * @param Tolerance The tolerance
+     * @param rBoundingBoxesWithTolerance The resulting bounding boxes with the applied tolerance
+     */
+    static void ComputeBoundingBoxesWithTolerance(
+        const std::vector<double>& rBoundingBoxes,
+        const double Tolerance,
+        std::vector<double>& rBoundingBoxesWithTolerance
+        );
+
+    /**
+     * @brief This method exchanges data asynchronously
+     * @tparam TDataType The type of data to be exchanged
+     * @param rSendBuffer The send buffer
+     * @param rRecvBuffer The receive buffer
+     * @param CommRank The rank of the current processor
+     * @param CommSize The size of the communicator
+     * @param rSendSizes The size of the send buffer
+     * @param rRecvSizes The size of the receive buffer
+     * @return 0 if the operation was successful
+     * @todo This method should be moved to the communicator
+     */
+    template<typename TDataType>
+    static int ExchangeDataAsync(
```

Comment:
this function should be part of the `DataCommunicator`

With a student I started on [this](https://github.com/KratosMultiphysics/Kratos/compare/master...core/mpi_async_communication) branch, but didnt have the time to finish it

Reply:
Huuum, okay